AvA scene group reworked

imPDA opened the pull request:
Hello! I would like to discuss compatibility with other addons in general and improvements to allianceWarSceneGroup in particular. I am working on UI addons and would like to make them work with PP seamlessly, so if any changes be done to PP, it would also affect my addon too, but I would also want to have more precise control (for example, be able to turn PP compatibility on and off). So all this leads to one conclusion - I should add compatibility to PP in my addon myself and not rely on compatibilty.lua inside PP.

It is pretty hard to do what I want now, let me explain on allianceWarSceneGroup example. I have added new (third tab) to this group, but PP knows nothing about it and my scene stays unchanged. I want to change it in the same manner as PP changed first two tabs (overview and browser), but how can I do it? The only way is to copy part of the code from PP and paste to my addon. Then I should check, if PP is enabled and invoke or not the code from PP. But if PP changes in the future, my code will possibly break and I will need to copy updated functions. 

After some changes the same can be done like this (snippet from my wip addon):
```lua
    local function MakeItPerfect()
        PP.allianceWarSceneGroupMainStuff(MY_NEW_SCENE, topLevelControl)
    end

    if PP then
        ZO_PostHook(PP, 'allianceWarSceneGroup', MakeItPerfect)
    end
```
So I only need to hook a function which changes two first tabs (but MakeItPerfect can invoked in any different way) and call the same function from PP, which will change MY_NEW_SCENE in the same manner as Overview and Browser were changed. As a result, I will receive good-looking tab without struggling from copy-pasting. And this is true for other menus - many authors could make their addons compatible if PP had more standardized ways of doing stuff.

I made changes to PerfectPixel\scenes\allianceWarSceneGroup.lua, please check it out. Please note I made variables easy to understand what they should be and them should be changed afterwards.

imPDA committed:
AvA sceen group reworked

DakJaniels commented:
You should be able to do what this person did for journal quest log, without needing to have pp make any code changes. But since you already have a example maybe it can be added. https://www.esoui.com/downloads/info3662-PerfectPixel-JournalQuestLog.html#info

imPDA committed:
journalSceneGroup updated

imPDA commented:
Well, this addon is an example what I want to avoid - copy-pasting the code, especially from another addons. Imagine you want to change something inside PP, lets say, this line
```lua
PP:CreateBackground(tlc, --[[#1]] nil, nil, nil, -10, -10, --[[#2]] nil, nil, nil, 0, 10)
```
to this
```lua
PP:CreateBackground(tlc, --[[#1]] nil, nil, nil, -10, -15, --[[#2]] nil, nil, nil, 0, 10)
```
(y offset changed), and now background of all default tabs 5px higher except JQL, because it has this line manually copy-pasted, and in order to match the new version of PP, it need to be updated too, and ALL other addons which want to be compatible with PP now have to do so. And PerfectPixel-JournalQuestLog not even maintained rn, and the author of this addon gave up trying to fix missing blur:

[FoG](https://www.esoui.com/forums/member.php?u=74424)
_I looked into the blur effect, You are correct and I originally thought it was a problem with Journal Quest Log, but after disabling my Addon it seems it also blurs. I did look at the code to see if I could figure out what I'm doing to remove the blur but cannot find it._

So I have done the same changes to journalSceneGroup.lua of PP. Now all you need to make JQL compatible with PP is
```lua
local function MakeItPerfect()
    PP.journalSceneGroupMainStuff(JQL_SCENE, JQL_Window)
end

if PP then
    ZO_PostHook(JournalQuestLog, 'Initialize', MakeItPerfect)
end
```
(and it even has the missing blur effect now). You actually can add it to PP itself to make JQL compatible or JQL author can add it to their addon (preferrable IMO), or it would be even better to make PP functions idempotent to avoid possible troubles if this piece of code will be added both in PP and in JQL just in case.

 
![image](https://github.com/user-attachments/assets/ced8191d-29ca-448a-8d75-6984d14b776b)

imPDA commented:
I did nothing besides decomposition of the initial code to logical units for better understanding and extraction of one function to be able to use it in other addons. It can be done to all scenes so PP can be actually compatible with any addon, you just need to provide a simple instruction in description :) So this code is fully back-compatible with everything done so far, but it opens new opportunities for other authors.

I can contribute and gradually make this changes to PP (at least for scenes where it is actually needed).

imPDA commented:
Just tested it with ESO Profiler, "it just works". So if you have any requests for making any other addon compatible with PP in mind - it can be done in 6 rows of code (or less if compact it a bit). If this PR will be reviewed and approved :)

```lua
local function MakeItPerfect2()
    PP.journalSceneGroupMainStuff(ESO_PROFILER_SCENE, ESOProfilerTopLevel)
end

if PP then
    ZO_PostHook(ESO_PROFILER, 'Initialize', MakeItPerfect2)
end
```
![image](https://github.com/user-attachments/assets/aac3da90-b10a-40c0-b3de-f2cee91b77b1)

Baertram commented:
Thanks for your PR, will need some time to check it and see what the actual difference is.

@imPDA So what exactly is the difference for the Alliance War Scene now?
I can see your changed code for the function
```
PP.allianceWarSceneGroup = function()
	for _, scene in ipairs(DEFAULT_ALLIANCE_WAR_SCENES) do
		MainStuffMustDoneHere(scene.sceneObj, scene.someGlobalObj.control)
	end

	AdditionalStuff()
end
```

But all it does is the same as it did before, only split up in mainstuff and additionalstuff functions? You got no new params added to PP.allianceWarSceneGroup!
-> Your example aboved showed new parameters: PP.allianceWarSceneGroupMainStuff(MY_NEW_SCENE, topLevelControl)

Could you please review your PR and check if you uploaded the correct code, and change the "*Stuff" function names to maybe
TLC_Changes() -> main
AdditionalChanges()

Thanks for reviewing it

imPDA committed:
rafactored: renamed

imPDA commented:
Yes, I did exactly this
> only split up in mainstuff and additionalstuff functions
and made "mainstuff" function available from outside via PP.allianceWarSceneGroup**MainStuff** function.

I have renamed functions, lets say, for each scene group (allianceWar, journal, etc.) we want to do two things: 
- change scene as whole by removing particular fragments and adding background
- change some elements (their positions) because PP changing anchors of topLevelControl at the first step (usually)
So lets name the first step ChangeScene and the second - ChangeElements.

![image](https://github.com/user-attachments/assets/557f2dcc-cefd-4371-bd81-d4cd6c39ad25)
(the same was done to journalSceneGroup, but it contains more code, so I split it a bit more, but the structure is the same)

The main change - at string 23(on the right part). This will allow to change all custom scenes in the same manner, as I demonstrated previously with JQL and ESOProfiler, like this:

```lua
local function MakeJQLPerfect()
    PP.journalSceneGroupEditScene(JQL_SCENE, JQL_Window)
end

if PP then
    ZO_PostHook(JournalQuestLog, 'Initialize', MakeJQLPerfect)
end

local function MakeESOProfilerPerfect()
    PP.journalSceneGroupEditScene(ESO_PROFILER_SCENE, ESOProfilerTopLevel)
end

if PP then
    ZO_PostHook(ESO_PROFILER, 'Initialize', MakeESOProfilerPerfect)
end
```

You can reach me out in Discord, it will make discussion faster and after I post main discussed points here in PR to keep track of things.

imPDA commented:
> all it does is the same as it did before

And this is second important point, I don't want to break any third-parties addons utilizing PP :)

Baertram commented:
Thanks for the details, I missed that "MainStuff" suffix at the examples and wondered where the params at the old functions were added :-)
Makese sense now.

>So lets name the first step ChangeScene and the second - ChangeElements.

Alright with me

But it still misses the other changes applied, like scrollbars etc.
I'll work on the functions PP.journalSceneGroupEditScene and add an optional 3rd parameter "otherElementsFunction" where you can pass in additional changes to your addon's other controls, like list re-anchor, scrollbar in PP style etc.

I do not really use Discord, dislike it.

Edit:
I've updated PP main branch with your code idea and added the 3rd param: editElementsFunc, for testing.

btw I tried your code with the ESO profiler and for me it did not apply the PP background properly/automatically.
ESO_PROFILER is not applying the backdrop and PP style automatically with that ZO_PostHook(ESO_PROFILER, 'Initialize', function() in compatibility.lua
I think the problem is that ESO_PROFILER.Initialize is called before PP loads and thus the hook never get's executed.
->>> We should add this code to the scenes:OnShow or Fragments:OnShow as they first show (if they got a scene or fragment), or use the OnDeferredInitialization function PP provides (if that is a ZO_DeferredInitializingObject)!

And due to that fact (timing) your idea won't work for these scenes, unless you call the code from your own addon which always loads before the other addons (and therefor you need to pass your addon name to the ## OptionalDependsOn: of that other addon, here: ESO Profiler).

This works (but the backdrop is hidden then until you switch away from ESO profiler tab adn back to it...
```
-- ==ESO_PROFILER==
        if ESO_PROFILER then
            local function applyElementChanges()

            end
            local esoProfilerChanged = false
            local scene = ESO_PROFILER_SCENE
            scene:RegisterCallback("StateChange", function(oldState, newState)
                if newState == SCENE_SHOWN and not esoProfilerChanged then
d("[PP]ESO_PROFILER init")
                    PP.journalSceneGroupEditScene(scene, ESOProfilerTopLevel, applyElementChanges)
                    esoProfilerChanged = true
                end
            end)
        end
```


I always wanted to add some PP.sceneStateChangeCallback functions, so maybe it's time for that now.

Baertram commented:
I've updated PP main branch with your code idea and added the 3rd param: editElementsFunc, for testing.
Also added a new function PP.onStateChangeCallback(sceneOrFragment, callbackFunc)

Implemented your code and directly called it from compatibility.lua of PP.
ESO_Profiler and QuestJournalLog work now.

1 thing I wonder is:
The new created PP backdrop, via function PP.journalSceneGroupEditScene, is always hidden on first open.
Happens because of PP:CreateBackground using this:
```
else
		bg		= CreateControl(parent:GetName() .. "_PP_BG", self.TLW_BG, CT_BACKDROP)
		bg:SetHidden(true)
```

Usually this get's unhidden as the scene shows the fragment. If we move the code to SCENE_SHOWING (earlier) it though misses some controls :-( So all I can do for now is create it at SCENE_SHOWN state and then unhide the new created backdrop once again.

imPDA commented:
So this is why I mentioned 

> JQL author can add it to their addon (preferrable IMO)

(or author of any other addon in general) because you can't take into account all details of every addon. But what you can do is to provide some API for other developers to wrap UI elements.

One more problem is dependency circling, you cant make another addon compatible with PP both from PP itself and from another addon at the same time. Lets take "X" addon as an example. You should specify OptionalDependsOn: X in PP and OptionalDependsOn: PerfectPixel in X. I dont know if they will be loaded correctly and in which order (but I am sure you will receive a warning form UI about circle dependencies). 

I think it would be better to move all this stuff to additional addon (PerfectPixelCompatibilities?) which will be depend on PP and optionally depend on all addons you want to make compatible with. This way you avoid circular dependency and clean PP a bit. I mean, it is logical step to split responsibilities between two addons IMO, and you can update PP and "compatibility" addon independently. I have not daven deep into PP structure yet, but I think it is pure structure-related changes and it can be done by copy-pasting all code from compatibilities.lua to new addon and adding simple OnAddonLoad initialization. It is also my thoughts about what can go wrong in the future from architectural point of view :D It will never probably be the case :D

Going to return to my addon now and test new functionality!

imPDA commented:
P.S. While I was writing this comment I set up the barebone of "PerfectPixelCompatibilities", take a look at [PerfectPixelCompatibilities](https://github.com/imPDA/PerfectPixelCompatibility). It mostly works as expected and should have no (or at least) less issues than setting up it inside of PP. It loads perfectly fine for me, except some bar. It need further investigation, but I believe if I you only change the background and anchors with `journalSceneGroupEditScene`  without changing small elements - it is fine, it will mimic the PerfectPixel enough and further improvements can take too much time and it is simply not worthy :)

Baertram commented:
>One more problem is dependency circling

Yes it definately was a problem already! You also get circular dependencies via libraries used in the addons. Thus PP is not adding any dependencies where not needed anymore and instead we use the proper scene, fragments, init or deferred init functions to load the compatibility code.

> PerfectPixelCompatibilities

This was also attempted in the past and in the end the same problems occured. You can always try to setup something like this and it works as long as it is maintained. But there is actually no difference doing this in the 1 addon or directly in PP in the end.
Doing it directly within PP often solved many other issues as you could react on other compatibility in time/more easily.


**In the end this addon PP here is not mine! Please ask @KL1SK what he thinks about it!** For me this is a project in a project and I got no time to wrok on this, sorry. I'm only maintaining PP and keeping it stable.

I'm fine with adding that extra functions for the journal scene and alliance war scene, as I did it in my current version here, and provide credit to you for sure.
If the support for eso profiler and quest journal are within PP or should be loaded from outside (their addons): I do not actually care. The current code "from within PP" works fine and I'd prefer that for now.

Baertram commented:
Added the tested functions to the PP files. Thanks for the ideas
